File: 03-2.py
#!/usr/bin/env python

from collections import namedtuple


class Number:
    def __init__(self, x, y):
        self.left = x
        self.right = x
        self.up = y
        self.down = y
        self.value = 0

    def add_digit(self, digit, x, y):
        self.value = self.value * 10 + int(digit)
        self.right = x

    def is_adjacent(self, pos):
        return pos.x >= self.left - 1 and \
            pos.x <= self.right + 1 and \
            pos.y >= self.up - 1 and \
            pos.y <= self.down + 1

    def __str__(self):
        return f"<{self.__class__.__name__}: {self.value}, l: {self.left}, r: {self.right}, u: {self.up}, d:{self.down}>"


class Gear:
    def __init__(self, pos):
        self.pos = pos
        self.parts = []

    def add_part(self, num):
        self.parts.append(num)

    def is_valid(self):
        return len(self.parts) == 2

    def ratio(self):
        return self.parts[0].value * self.parts[1].value

    def __str__(self):
        return f"<{self.__class__.__name__}: valid: {self.is_valid()}, x: {self.pos.x}, y: {self.pos.y}, parts: {self.parts}>"


input = [row.strip() for row in open('03.input').readlines()]

numbers = []
symbols = []
gears = []

for y in range(len(input)):
    current_number = None
    row = input[y]
    for x in range(len(row)):
        char = row[x]
        match char:
            case '.':
                current_number = None
            case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9':
                if not current_number:
                    current_number = Number(x, y)
                    numbers.append(current_number)
                current_number.add_digit(char, x, y)
            case _:
                current_number = None
                pos = namedtuple('Position', ['x', 'y'])
                pos.x = x
                pos.y = y
                symbols.append(pos)
                if char == '*':
                    gears.append(Gear(pos))

for gear in gears:
    for num in numbers:
        if num.is_adjacent(gear.pos):
            gear.add_part(num)

print(sum([g.ratio() for g in gears if g.is_valid()]))

File: 03-2.rb
#!/usr/bin/env ruby

require 'ostruct'

class Number
  attr_accessor :left, :right, :up, :down, :value

  def initialize(x, y)
    @left = x
    @right = x
    @up = y
    @down = y
    @value = 0
  end

  def add_digit(digit, x, y)
    @value = @value * 10 + digit.to_i
    @right = x
  end

  def adjacent?(pos)
    pos.x >= @left - 1 and
      pos.x <= @right + 1 and
      pos.y >= @up - 1 and
      pos.y <= @down + 1
  end

  def to_s
    "<#{self.class}: #{@value}, l: #{@left}, r: #{@right}, u: #{@up}, d: #{@down}>"
  end
end

class Gear
  attr_accessor :parts, :pos

  def initialize(pos)
    @pos = pos
    @parts = []
  end

  def add_part(num)
    @parts << num
  end

  def valid?
    @parts.size == 2
  end

  def ratio
    @parts[0].value * @parts[1].value
  end

  def to_s
    "<#{self.class}: valid: #{valid?}, x: #{@pos.x}, y: #{@pos.y}, parts: #{@parts}>"
  end
end

input = File.read('03.input').lines.map(&:strip).map(&:chars)

numbers = []
symbols = []
gears = []

input.each_with_index do |row, y|
  current_number = nil
  row.each_with_index do |char, x|
    case char
    when '.'
      current_number = nil
    when '0'..'9'
      unless current_number
        current_number = Number.new(x, y)
        numbers << current_number
      end
      current_number.add_digit(char, x, y)
    else
      current_number = nil
      pos = OpenStruct.new
      pos.x = x
      pos.y = y
      symbols << pos
      gears << Gear.new(pos) if char == '*'
    end
  end
end

gears.each do |gear|
  numbers.each do |num|
    gear.add_part num if num.adjacent? gear.pos
  end
end

print gears.select(&:valid?).map(&:ratio).sum, "\n"
